File: src/scripts/lib/catchup.ts
import { fmtBase } from "@/scripts/lib/creds.js"

export interface CatchupChannel {
  id: number | string
  url?: string | null
  catchup?: string | null
  catchupDays?: number | null
  catchupSource?: string | null
}

export interface CatchupProgramme {
  start: number
  stop: number
}

export interface CatchupCreds {
  host?: string
  port?: string
  user?: string
  pass?: string
  liveContainer?: string | null
}

const DEFAULT_CATCHUP_DAYS = 7

function preferredLiveContainer(creds: CatchupCreds): "hls" | "ts" {
  const configured = String(creds.liveContainer || "").trim().toLowerCase()
  if (configured === "ts" || configured === "mpegts") return "ts"
  if (configured === "hls" || configured === "m3u8") return "hls"
  return "hls"
}

function seconds(ms: number): number {
  return Math.floor(ms / 1000)
}

function durationMinutes(programme: CatchupProgramme): number {
  return Math.max(1, Math.ceil((programme.stop - programme.start) / 60_000))
}

function formatXtreamStart(ts: number): string {
  const d = new Date(ts)
  const pad = (value: number) => String(value).padStart(2, "0")
  return (
    `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}:` +
    `${pad(d.getHours())}-${pad(d.getMinutes())}`
  )
}

function appendQuery(url: string, query: string): string {
  return `${url}${url.includes("?") ? "&" : "?"}${query}`
}

function replaceCatchupPlaceholders(
  template: string,
  programme: CatchupProgramme,
): string {
  const start = seconds(programme.start)
  const stop = seconds(programme.stop)
  const duration = Math.max(1, stop - start)
  const minutes = durationMinutes(programme)
  const now = seconds(Date.now())
  const replacements: Record<string, string | number> = {
    start,
    end: stop,
    stop,
    utc: start,
    lutc: now,
    timestamp: now,
    duration,
    "duration-minutes": minutes,
    offset: start,
  }
  return template.replace(/\$\{([^}]+)\}|\{([^}]+)\}/g, (match, a, b) => {
    const key = String(a || b || "").trim().toLowerCase()
    return replacements[key] == null ? match : String(replacements[key])
  })
}

export function channelHasCatchup(channel: CatchupChannel | null | undefined): boolean {
  if (!channel) return false
  const mode = String(channel.catchup || "").trim().toLowerCase()
  return (
    !!channel.catchupSource ||
    !!channel.catchupDays ||
    mode === "xtream" ||
    mode === "append" ||
    mode === "default" ||
    mode === "shift" ||
    mode === "flussonic"
  )
}

export function canReplayProgramme(
  channel: CatchupChannel | null | undefined,
  programme: CatchupProgramme | null | undefined,
  now = Date.now(),
): boolean {
  if (!channelHasCatchup(channel) || !programme) return false
  if (!Number.isFinite(programme.start) || !Number.isFinite(programme.stop)) return false
  if (programme.stop > now || programme.stop <= programme.start) return false
  const days = Number(channel?.catchupDays)
  const windowDays = Number.isFinite(days) && days > 0 ? days : DEFAULT_CATCHUP_DAYS
  return programme.start >= now - windowDays * 24 * 60 * 60 * 1000
}

export function buildCatchupStreamUrl(
  channel: CatchupChannel,
  programme: CatchupProgramme,
  creds: CatchupCreds = {},
): string | null {
  if (!canReplayProgramme(channel, programme)) return null
  if (channel.catchupSource) {
    return replaceCatchupPlaceholders(channel.catchupSource, programme)
  }

  const mode = String(channel.catchup || "").trim().toLowerCase()
  if (channel.url && (mode === "append" || mode === "default" || mode === "shift" || mode === "flussonic" || !!channel.catchupDays)) {
    return appendQuery(
      channel.url,
      `utc=${seconds(programme.start)}&lutc=${seconds(Date.now())}`,
    )
  }

  if (mode === "xtream" && creds.host && creds.user && creds.pass) {
    const ext = preferredLiveContainer(creds) === "ts" ? ".ts" : ".m3u8"
    return (
      fmtBase(creds.host, creds.port || "") +
      "/timeshift/" +
      encodeURIComponent(creds.user) +
      "/" +
      encodeURIComponent(creds.pass) +
      "/" +
      durationMinutes(programme) +
      "/" +
      encodeURIComponent(formatXtreamStart(programme.start)) +
      "/" +
      encodeURIComponent(String(channel.id)) +
      ext
    )
  }

  return null
}

File: native/flutter/leleg_iptv/lib/domain/catchup.dart
import 'xtream_client.dart';

/// Catch-up / timeshift URL builders aligned with `src/scripts/lib/catchup.ts`.
class Catchup {
  static const defaultCatchupDays = 7;

  static bool channelHasCatchup(LiveChannel channel) {
    final mode = channel.catchup.trim().toLowerCase();
    return channel.catchupSource.isNotEmpty ||
        channel.catchupDays > 0 ||
        mode == 'xtream' ||
        mode == 'append' ||
        mode == 'default' ||
        mode == 'shift' ||
        mode == 'flussonic';
  }

  static bool canReplayProgramme(
    LiveChannel? channel,
    EpgProgramme programme, {
    DateTime? now,
  }) {
    final clock = now ?? DateTime.now();
    if (channel == null || !channelHasCatchup(channel)) return false;
    final start = programme.start;
    final end = programme.end;
    if (start == null || end == null) return false;
    if (end.isAfter(clock) || !end.isAfter(start)) return false;
    final windowDays = channel.catchupDays > 0
        ? channel.catchupDays
        : (channelHasCatchup(channel) ? defaultCatchupDays : 0);
    if (windowDays <= 0) return false;
    return start.isAfter(clock.subtract(Duration(days: windowDays)));
  }

  static List<String> buildStreamUrls(
    XtreamProfile profile,
    LiveChannel channel,
    EpgProgramme programme, {
    required String Function(LiveChannel channel) liveUrl,
    DateTime? now,
  }) {
    final clock = now ?? DateTime.now();
    if (!canReplayProgramme(channel, programme, now: clock)) {
      return const [];
    }
    final start = programme.start;
    final end = programme.end;
    if (start == null || end == null) return const [];

    final urls = <String>[];
    final mode = channel.catchup.trim().toLowerCase();

    final source = channel.catchupSource.trim();
    if (source.isNotEmpty) {
      urls.add(_replaceCatchupPlaceholders(source, start, end, clock));
    }

    if (mode == 'xtream' ||
        channel.catchupDays > 0 ||
        (mode.isEmpty && channelHasCatchup(channel))) {
      urls.addAll(_xtreamTimeshiftUrls(profile, channel, programme));
    }

    final liveBase = channel.directSource.trim().isNotEmpty
        ? channel.directSource.trim()
        : liveUrl(channel);
    if (liveBase.isNotEmpty) {
      final appendAllowed =
          mode == 'append' ||
          mode == 'default' ||
          mode == 'shift' ||
          mode == 'flussonic' ||
          (channel.catchupDays > 0 && mode != 'xtream');
      if (appendAllowed) {
        final startSec = start.millisecondsSinceEpoch ~/ 1000;
        final stopSec = end.millisecondsSinceEpoch ~/ 1000;
        final nowSec = clock.millisecondsSinceEpoch ~/ 1000;
        final separator = liveBase.contains('?') ? '&' : '?';
        urls.add('$liveBase${separator}utc=$startSec&lutc=$nowSec');
        urls.add(
          '$liveBase${separator}utc=$startSec&lutc=$nowSec&duration=${stopSec - startSec}',
        );
      }
    }

    return urls.toSet().toList(growable: false);
  }

  static List<String> _xtreamTimeshiftUrls(
    XtreamProfile profile,
    LiveChannel channel,
    EpgProgramme programme,
  ) {
    final start = programme.start;
    final end = programme.end;
    if (start == null || end == null) return const [];
    final durationMinutes = (end.difference(start).inSeconds / 60).ceil();
    final safeDuration = durationMinutes < 1 ? 1 : durationMinutes;
    final stamp = _formatXtreamStart(start);
    final primary = profile.liveContainer == 'ts' ? 'ts' : 'm3u8';
    final alternate = primary == 'ts' ? 'm3u8' : 'ts';
    final user = profile.username;
    final pass = profile.password;
    final plainBase =
        '${profile.baseUrl}/timeshift/$user/$pass/$safeDuration/$stamp/${channel.id}';
    final encodedBase =
        '${profile.baseUrl}/timeshift/'
        '${Uri.encodeComponent(user)}/'
        '${Uri.encodeComponent(pass)}/'
        '$safeDuration/'
        '${Uri.encodeComponent(stamp)}/'
        '${Uri.encodeComponent(channel.id.toString())}';
    return [
      '$encodedBase.$primary',
      '$plainBase.$primary',
      '$encodedBase.$alternate',
      '$plainBase.$alternate',
    ];
  }

  static String _formatXtreamStart(DateTime value) {
    String pad(int number) => number.toString().padLeft(2, '0');
    return '${value.year}-${pad(value.month)}-${pad(value.day)}:'
        '${pad(value.hour)}-${pad(value.minute)}';
  }

  static String _replaceCatchupPlaceholders(
    String template,
    DateTime start,
    DateTime end,
    DateTime now,
  ) {
    final startSec = start.millisecondsSinceEpoch ~/ 1000;
    final stopSec = end.millisecondsSinceEpoch ~/ 1000;
    final durationSec = stopSec - startSec < 1 ? 1 : stopSec - startSec;
    final durationMinutes = (durationSec / 60).ceil().clamp(1, 99999);
    final nowSec = now.millisecondsSinceEpoch ~/ 1000;
    final replacements = <String, String>{
      'start': '$startSec',
      'end': '$stopSec',
      'stop': '$stopSec',
      'utc': '$startSec',
      'lutc': '$nowSec',
      'timestamp': '$nowSec',
      'duration': '$durationSec',
      'duration-minutes': '$durationMinutes',
      'offset': '$startSec',
    };
    return template.replaceAllMapped(RegExp(r'\$\{([^}]+)\}|\{([^}]+)\}'), (
      match,
    ) {
      final key = (match.group(1) ?? match.group(2) ?? '').trim().toLowerCase();
      return replacements[key] ?? match.group(0)!;
    });
  }
}
